IPython/core/history.py

```diff
@@ -81,18 +80,29 @@ def catch_corrupt_db(f, self, *a, **kw):
     try:
         return f(self, *a, **kw)
     except (DatabaseError, OperationalError) as e:
-        if os.path.isfile(self.hist_file):
-            # Try to move the file out of the way
-            base,ext = os.path.splitext(self.hist_file)
-            newpath = base + '-corrupt' + ext
-            os.rename(self.hist_file, newpath)
-            print("ERROR! History file wasn't a valid SQLite database (%s)." % e,
-            "It was moved to %s" % newpath, "and a new file created.", file=sys.stderr)
+        self._corrupt_db_counter += 1
+        self.log.error("Failed to open SQLite history %s (%s).", self.hist_file, e)
+        if self.hist_file != ':memory:':
+            if self._corrupt_db_counter > self._corrupt_db_limit:
+                self.hist_file = ':memory:'
+                self.log.error("Failed to load history too many times, history will not be saved.")
+            elif os.path.isfile(self.hist_file):
+                # Try to move the file out of the way
+                base,ext = os.path.splitext(self.hist_file)
+                now = datetime.datetime.now().isoformat().replace(':', '.')
+                newpath = base + '-corrupt-' + now + ext
```

Comment:
I appreciate the reason for this, but I'm not sure I like the solution - in a case like #9412, if the user doesn't configure a different location, it's presumably going to create two new, empty, corrupt database files for each IPython session, until the profile directory is littered with thousands of them.

Maybe it could just include the corrupt_db_counter - that way you preserve all the files from the last attempt to start IPython, but they're not accumulating.